src/server/client.rs

```diff
@@ -114,19 +115,56 @@ pub async fn client_rpc_loop<
 }
 
 async fn handle_worker_stop(
+    state_ref: &StateRef,
     tako_ref: &TakoServer,
-    worker_id: WorkerId,
-) -> crate::Result<ToClientMessage> {
-    match tako_ref
-        .send_message(FromGatewayMessage::StopWorker(StopWorkerRequest {
-            worker_id,
-        }))
-        .await?
-    {
-        ToGatewayMessage::WorkerStopped => Ok(ToClientMessage::StopWorkerResponse),
-        ToGatewayMessage::Error(error) => Ok(ToClientMessage::Error(error.message)),
-        msg => panic!("Received invalid response to worker stop: {:?}", msg),
+    selector: WorkerSelector,
+) -> ToClientMessage {
+    let state = state_ref.get();
+    let mut responses: Vec<(WorkerId, StopWorkerResponse)> = Vec::new();
+
+    let worker_ids: Vec<WorkerId> = match selector {
+        WorkerSelector::Specific(ids) => ids,
+        WorkerSelector::All => state
+            .get_workers()
+            .iter()
+            .filter(|(_, worker)| worker.make_info().ended.is_none())
+            .map(|(_, worker)| worker.worker_id())
+            .collect(),
+    };
+
+    for worker_id in worker_ids {
+        if let Some(worker) = state.get_worker(worker_id) {
+            if worker.make_info().ended.is_some() {
+                responses.push((
+                    worker_id,
+                    StopWorkerResponse::Failed("worker is already stopped".to_string()),
```

Comment:
IMO this shouldn't be an error. At least it should be renamed to something like `AlreadyStopped`.